File: app/Models/Transaction.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Transaction extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'type',        // usage, reward, purchase, refund
        'action_type', // quiz_generation, flashcard_generation, scan_solve, ai_grading, referral_bonus, signup_bonus
        'amount',      // integer (positive for credits added, negative for credits used)
        'description',
        'model_used',  // claude-sonnet-4-5, deepseek-chat, etc.
        'request_id',  // idempotency / correlation key
        'metadata',
    ];

    protected $casts = [
        'metadata' => 'array',
        'amount' => 'integer',
    ];

    /**
     * Get the user that owns the transaction.
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }
}

File: skeeme-go/internal/models/transaction.go
package models

import (
	"database/sql"
)

type Transaction struct {
	ID          int64          `db:"id" json:"id"`
	UserID      int64          `db:"user_id" json:"user_id"`
	Type        string         `db:"type" json:"type"` // e.g. "deduction", "refund", "purchase"
	ActionType  sql.NullString `db:"action_type" json:"action_type"`
	ModelUsed   sql.NullString `db:"model_used" json:"model_used"`
	RequestID   sql.NullString `db:"request_id" json:"request_id"`
	Amount      int            `db:"amount" json:"amount"` // Positive = credit, Negative = debit
	Description sql.NullString `db:"description" json:"description"`
	Metadata    []byte         `db:"metadata" json:"metadata"` // JSONB
	CreatedAt   sql.NullTime   `db:"created_at" json:"created_at"`
	UpdatedAt   sql.NullTime   `db:"updated_at" json:"updated_at"`
}
